LUSDChickenBonds/src/BLUSDLPZap.sol

```diff
@@ -0,0 +1,95 @@
+
+    IERC20 constant public lusdToken = IERC20(LUSD_TOKEN_ADDRESS);
+    IERC20 constant public bLUSDToken = IERC20(BLUSD_TOKEN_ADDRESS);
+    ICurvePool constant public lusd3CRVPool = ICurvePool(LUSD_3CRV_POOL_ADDRESS);
+    ICurveCryptoPool constant public bLUSDLUSD3CRVPool = ICurveCryptoPool(BLUSD_LUSD_3CRV_POOL_ADDRESS);
+    IERC20 constant public bLUSDLUSD3CRVLPToken = IERC20(BLUSD_LUSD_3CRV_LP_TOKEN_ADDRESS);
+    ICurveLiquidityGaugeV5 constant public bLUSDGauge = ICurveLiquidityGaugeV5(BLUSD_LUSD_3CRV_GAUGE_ADDRESS);
+
+    // TODO: add permit version
+    function _addLiquidity(
+        uint256 _bLUSDAmount,
+        uint256 _lusdAmount,
+        uint256 _minLPTokens,
+        address _receiver
+    )
+        internal
+        returns (uint256 bLUSDLUSD3CRVTokens)
+    {
+        require(_bLUSDAmount > 0 || _lusdAmount > 0, "BLUSDLPZap: Cannot provide zero liquidity");
+
+        uint256 lusd3CRVAmount;
+        if (_lusdAmount > 0) {
+            lusdToken.transferFrom(msg.sender, address(this), _lusdAmount);
+
+            // add LUSD single sided
+            lusdToken.approve(address(lusd3CRVPool), _lusdAmount);
+            lusd3CRVAmount = lusd3CRVPool.add_liquidity([_lusdAmount, 0], 0, address(this));
+        }
+
+        bLUSDToken.transferFrom(msg.sender, address(this), _bLUSDAmount);
+
+        // add bLUSD/LUSD-3CRV
+        bLUSDToken.approve(address(bLUSDLUSD3CRVPool), _bLUSDAmount);
+        if (lusd3CRVAmount > 0) {
+            lusd3CRVPool.approve(address(bLUSDLUSD3CRVPool), lusd3CRVAmount);
+        }
+        bLUSDLUSD3CRVTokens = bLUSDLUSD3CRVPool.add_liquidity([_bLUSDAmount, lusd3CRVAmount], _minLPTokens, false, _receiver);
+
+        return bLUSDLUSD3CRVTokens;
+    }
+
+    function addLiquidity(uint256 _bLUSDAmount, uint256 _lusdAmount, uint256 _minLPTokens) external returns (uint256 bLUSDLUSD3CRVTokens) {
+        // add liquidity
+        bLUSDLUSD3CRVTokens = _addLiquidity(_bLUSDAmount, _lusdAmount, _minLPTokens, msg.sender);
+
+        return bLUSDLUSD3CRVTokens;
+    }
+
+    function addLiquidityAndStake(
+        uint256 _bLUSDAmount,
+        uint256 _lusdAmount,
+        uint256 _minLPTokens
+    )
+        external
+        returns (uint256 bLUSDLUSD3CRVTokens)
+    {
+        // add liquidity
+        bLUSDLUSD3CRVTokens = _addLiquidity(_bLUSDAmount, _lusdAmount, _minLPTokens, address(this));
+
+        // approve LP tokens to Gauge
+        bLUSDLUSD3CRVLPToken.approve(address(bLUSDGauge), bLUSDLUSD3CRVTokens);
+
+        // stake into gauge
+        bLUSDGauge.deposit(bLUSDLUSD3CRVTokens, msg.sender, false); // make sure rewards are not claimed
+
+        return bLUSDLUSD3CRVTokens;
+    }
+
+    function getMinLPTokens(uint256 _bLUSDAmount, uint256 _lusdAmount) external view returns (uint256 bLUSDLUSD3CRVTokens) {
+        uint256 lusd3CRVAmount;
+        if (_lusdAmount > 0) {
+            lusd3CRVAmount = lusd3CRVPool.calc_token_amount([_lusdAmount, 0], true);
```

Comment:
StableSwap's `calc_token_amount()` is not designed to be precise, and it ignores fess. As such, if you use it to calculate a minimum expected LP token amount, it'll be higher than what you'd normally get, and your TX will fail. This explains the TODO in the corresponding test.

See inside `get_dy_underlying()` in here for a better approximation:
https://etherscan.io/address/0xEd279fDD11cA84bEef15AF5D39BB4d4bEE23F0cA#code

It's still not perfect (which is acknowledged in the code), so you still might need to lower it by some fraction to be usable in practice.

Reply:
Crypto (v2) `calc_token_amount()` should be accurate though.

Reply:
I was assuming that we would apply a slippage threshold that would be some orders of magnitude bigger than the inaccuracy here (~1% vs 0.04%), so it would still work, and erroring on the safe side for the user, but now I fixed it by adding the fees. Thanks again!